subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/NestedValidationUtil.java

```diff
@@ -51,33 +52,29 @@ public static void validateBeanType(
         String propertyName,
         Class<?> beanType
     ) {
-        if (!isSupportedType(beanType)) {
-            validationContext.visitPropertyProblem(problem ->
-                problem.withId(ValidationProblemId.NESTED_TYPE_UNSUPPORTED)
-                    .reportAs(WARNING)
-                    .forProperty(propertyName)
-                    .withDescription(() -> "with nested type '" + beanType.getName() + "' is not supported")
-                    .happensBecause("Nested types are expected to either declare some annotated properties or some behaviour that requires capturing the type as input")
-                    .addPossibleSolution("Declare a nested type, e.g. `Provider<T>`, `Iterable<T>`, or `<MapProperty<K, V>>`, where `T` and `V` have some annotated properties or some behaviour that requires capturing the type as input")
-                    .documentedAt("validation_problems", "unsupported_nested_type")
-            );
-        }
-    }
-
-    private static boolean isJavaSE(Class<?> type) {
-        return type.getName().startsWith("java.") || type.getName().startsWith("javax.");
-    }
-
-    private static boolean isKotlinStdlib(Class<?> type) {
-        return type.getName().startsWith("kotlin.") || type.getName().startsWith("kotlinx.");
+        Optional<String> unsupportedReason = isSupportedType(beanType);
+        unsupportedReason.ifPresent(reason -> validationContext.visitPropertyProblem(problem ->
+            problem.withId(ValidationProblemId.NESTED_TYPE_UNSUPPORTED)
+                .reportAs(WARNING)
+                .forProperty(propertyName)
+                .withDescription(() -> "with nested type '" + beanType.getName() + "' is not supported")
+                .happensBecause(reason)
+                .addPossibleSolution("Use a different input annotation if type is not a bean")
+                .addPossibleSolution("Use a different package that doesn't conflict with standard Java or Kotlin types for custom types")
+                .documentedAt("validation_problems", "unsupported_nested_type")
+        ));
     }
 
-    private static boolean isGString(Class<?> type) {
-        return type.getName().startsWith("groovy.lang.GString");
-    }
-
-    private static boolean isSupportedType(Class<?> type) {
-        return !isJavaSE(type) && !isKotlinStdlib(type) && !isGString(type);
+    private static Optional<String> isSupportedType(Class<?> type) {
```

Comment:
💅 Consider renaming this to something like `getUnsupportedReason`. The current name does not align well with the return type.

Reply:
Makes sense